# Company Name Matching
### Approach 
1. Preprocess to noramlize names: remove spaces, characters, lowercase, 
2. Build name embeddings using [StarSpace](https://github.com/facebookresearch/StarSpace)
3. Index Using [Annoy](https://github.com/spotify/annoy)
4. Evaluate Using accuracy on non -1 companies.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import FunctionTransformer
from sklearn.pipeline import make_pipeline
from sklearn import svm, linear_model
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import roc_curve, auc,f1_score
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from collections import defaultdict
import unicodedata
import random
random.seed(37)
import gensim
from gensim.models import Word2Vec
from annoy import AnnoyIndex

In [2]:
pd.set_option('display.max_colwidth', -1)

In [3]:
data_path = "data/"
file_path = "files/"

In [4]:
companies = pd.read_csv(data_path+"G.csv", delimiter="|")
train = pd.read_csv(data_path+"STrain.csv", delimiter="|")

In [5]:
companies.head()

company_id                        name
0  634022      PRIMCOM SA                
1  324497      The David Isaacs Fund     
2  280848      Bramor Enterprises Limited
3  432662      NAVEXIM S.A.              
4  524224      Magal Group SA

In [6]:
companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 450256 entries, 0 to 450255
Data columns (total 2 columns):
company_id    450256 non-null int64
name          450256 non-null object
dtypes: int64(1), object(1)
memory usage: 6.9+ MB


In [7]:
companies["company_id"].nunique()

450256

In [8]:
train.head()

train_index                                                          name  \
0  0            TRATTAMENTO Ltd RIFIUTI METROPOLITANI SPA  SIGLABILE TRM SPA   
1  1            A IRL Fuund                                                    
2  2            BMR-500 Kendall LLC 1 Mezz GmbH                                
3  3            Solich GmbH KG                                                 
4  4            Drzyzzga Funds Logi. sp. z oo                                  

   company_id  
0  177358      
1  568472      
2  195692      
3 -1           
4  404178

In [9]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 3 columns):
train_index    100000 non-null int64
name           100000 non-null object
company_id     100000 non-null int64
dtypes: int64(2), object(1)
memory usage: 2.3+ MB


In [10]:
print train.columns

Index([u'train_index', u'name', u'company_id'], dtype='object')


In [11]:
print train.shape
print train["name"].nunique()

(100000, 3)
99365


In [12]:
companies = companies.reset_index(drop=True)
companies.head()

company_id                        name
0  634022      PRIMCOM SA                
1  324497      The David Isaacs Fund     
2  280848      Bramor Enterprises Limited
3  432662      NAVEXIM S.A.              
4  524224      Magal Group SA

In [13]:
train.shape

(100000, 3)

In [14]:
train = train.reset_index(drop=True)
train.head()

train_index                                                          name  \
0  0            TRATTAMENTO Ltd RIFIUTI METROPOLITANI SPA  SIGLABILE TRM SPA   
1  1            A IRL Fuund                                                    
2  2            BMR-500 Kendall LLC 1 Mezz GmbH                                
3  3            Solich GmbH KG                                                 
4  4            Drzyzzga Funds Logi. sp. z oo                                  

   company_id  
0  177358      
1  568472      
2  195692      
3 -1           
4  404178

### Pre-process data

In [15]:
def remove_accents(df,**kw_args):
    old_col =  kw_args["old_col"]
    new_col =  kw_args["new_col"]
    
    def remove_accents_inner(input_str):
        try:
            nfkd_form = unicodedata.normalize('NFKD', unicode(input_str, 'utf8'))
        except UnicodeDecodeError:
#             print input_str
#             input_str = input_str.encode('utf-8')
#             nfkd_form = unicodedata.normalize('NFKD', unicode(input_str, 'utf8'))
            return None
        return u"".join([c for c in nfkd_form if not unicodedata.combining(c)])
    
    df[new_col] = df[old_col].apply(remove_accents_inner)
    return df

def make_accent_transformer(old_col, new_col):
    return FunctionTransformer(remove_accents, validate=False,
                                         kw_args={"old_col":old_col,"new_col":new_col})

In [16]:
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

[nltk_data] Downloading package stopwords to /home/akash/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [17]:
import re
REPLACE_BY_SPACE_RE = re.compile('[/(){}\[\]|@,;\-\\\+#~!$%^]')
BAD_SYMBOLS_RE = re.compile('[^0-9a-z ]')
STOPWORDS = set(stopwords.words('english'))
#-#+_!`~\\

In [18]:
s = "for from india and add+g-ddc.b.g,c\g-hh ak|ash#sin/gh dee$pt^a su%d"
s=re.sub(REPLACE_BY_SPACE_RE," ",s)
s= re.sub(BAD_SYMBOLS_RE,"",s)
s= ' '.join(word for word in s.split() if word not in STOPWORDS)
print s

india add g ddcbg c g hh ak ash sin gh dee pt su


In [19]:
def clean_name(df,**kw_args):
    old_col =  kw_args["old_col"]
    new_col =  kw_args["new_col"]
    
    def regex_clean(text):
        try:
            text = re.sub(REPLACE_BY_SPACE_RE, " ", text)
            text = re.sub(BAD_SYMBOLS_RE, "", text)
#         #add shorthand
#         longhand = ' '.join(word for word in text.split() if not any(c.isdigit() for c in word))
#         shorthand = "".join(word[0] for word in re.findall("\w+", longhand))
#         text = text+" "+shorthand
        except TypeError:
            print text
            
        return text
    
    def strip_extra_spaces(text):
        return " ".join(text.split())
    
    df[new_col] = df[old_col].str.lower().str.strip()
    df[new_col] = df[new_col].apply(regex_clean)
    df[new_col] = df[new_col].apply(strip_extra_spaces)
    
    return df

def make_clean_name_transformer(old_col, new_col):
    return FunctionTransformer(clean_name, validate=False, kw_args={"old_col":old_col,"new_col":new_col})

#### Make processing pipeline

In [20]:
name_pipeline = make_pipeline(make_accent_transformer("name", "clean_name"),
                                    make_clean_name_transformer("clean_name", "clean_name"))

#### Transform company names

In [21]:
companies = name_pipeline.transform(companies)
print companies.shape
print companies["clean_name"].nunique()

(450256, 3)
448050


#### Check top occurring words

In [22]:
from collections import Counter
count_vectorizer = CountVectorizer(analyzer='word', ngram_range=(1, 1), min_df=1)
# X matrix where the row represents sentences and column is our one-hot vector for each token in our vocabulary
X = count_vectorizer.fit_transform(companies["clean_name"])

# Vocabulary
vocab = list(count_vectorizer.get_feature_names())

# Column-wise sum of the X matrix.
# It's some crazy numpy syntax that looks horribly unpythonic
# For details, see http://stackoverflow.com/questions/3337301/numpy-matrix-to-array
# and http://stackoverflow.com/questions/13567345/how-to-calculate-the-sum-of-all-columns-of-a-2d-numpy-array-efficiently
counts = X.sum(axis=0).A1

freq_distribution = Counter(dict(zip(vocab, counts)))

In [23]:
print (freq_distribution.most_common(100))

[(u'llc', 44288), (u'fund', 38551), (u'limited', 34803), (u'gmbh', 29272), (u'inc', 24952), (u'bv', 21019), (u'srl', 20646), (u'trust', 20440), (u'ltd', 16548), (u'sa', 15202), (u'co', 12984), (u'lp', 11681), (u'global', 11272), (u'de', 10292), (u'the', 10220), (u'company', 10159), (u'funds', 9906), (u'kg', 9865), (u'holding', 9770), (u'ab', 9459), (u'investment', 8335), (u'sl', 7666), (u'capital', 7485), (u'international', 7283), (u'equity', 6886), (u'di', 6700), (u'holdings', 6683), (u'of', 6680), (u'spa', 6663), (u'group', 6212), (u'as', 6092), (u'ii', 6061), (u'sicav', 5883), (u'management', 5723), (u'investments', 5456), (u'aps', 5192), (u'portfolio', 4959), (u'bond', 4854), (u'partners', 4848), (u'societa', 4605), (u'invest', 4526), (u'and', 4211), (u'oy', 4152), (u'sro', 4143), (u'bank', 3981), (u'plc', 3861), (u'services', 3851), (u'income', 3848), (u'master', 3799), (u'corporation', 3747), (u'nv', 3573), (u'beheer', 3568), (u'mbh', 3509), (u'spoka', 3506), (u'ag', 3479), (u'se

#### Pre-process train

In [24]:
train = name_pipeline.transform(train)

In [25]:
train.head()

train_index                                                          name  \
0  0            TRATTAMENTO Ltd RIFIUTI METROPOLITANI SPA  SIGLABILE TRM SPA   
1  1            A IRL Fuund                                                    
2  2            BMR-500 Kendall LLC 1 Mezz GmbH                                
3  3            Solich GmbH KG                                                 
4  4            Drzyzzga Funds Logi. sp. z oo                                  

   company_id                                                   clean_name  
0  177358      trattamento ltd rifiuti metropolitani spa siglabile trm spa  
1  568472      a irl fuund                                                  
2  195692      bmr 500 kendall llc 1 mezz gmbh                              
3 -1           solich gmbh kg                                               
4  404178      drzyzzga funds logi sp z oo

In [26]:
# Settings for embeddings and indexing model
DIM = 100
NUM_TREES_ANNOY = 200
N_GRAMS =[3]

In [27]:
# function to make training name corpus using n grams
def make_corpus(names, n_grams):
    corpus = []
    for name in names:
        name_grams = []
        for n in n_grams:
            grams = [name[i:i+n] for i in range(len(name)-n+1)]
            name_grams.extend(grams)
        name_grams = [ gram.replace(" ","_") for gram in name_grams]
        corpus.append(" ".join(name_grams))
    return corpus

#### Make corpus of ngrams

In [28]:
company_corpus =  make_corpus(companies["clean_name"], N_GRAMS)

In [29]:
companies["n_grams"] = company_corpus
companies

company_id  \
0       634022       
1       324497       
2       280848       
3       432662       
4       524224       
5       513585       
6       354496       
7       381944       
8       526057       
9       34381        
10      170089       
11      425961       
12      606644       
13      449644       
14      85335        
15      338841       
16      280937       
17      622744       
18      260264       
19      561089       
20      632695       
21      103863       
22      256892       
23      133701       
24      80822        
25      637249       
26      596390       
27      366252       
28      140358       
29      115667       
...        ...       
450226  348330       
450227  510306       
450228  334917       
450229  540797       
450230  557856       
450231  127359       
450232  549956       
450233  350077       
450234  302899       
450235  255907       
450236  239536       
450237  482822       
450238  70003        
450239  283503       
450240  4804         
450241  219015       
450242  452876       
450243  316806       
450244  428951       
450245  478239       
450246  554608       
450247  497982       
450248  37736        
450249  369911       
450250  477477       
450251  614780       
450252  26499        
450253  581457       
450254  401977       
450255  497444       

                                                                                                                                      name  \
0       PRIMCOM SA                                                                                                                           
1       The David Isaacs Fund                                                                                                                
2       Bramor Enterprises Limited                                                                                                           
3       NAVEXIM S.A.                                                                                                                         
4       Magal Group SA                                                                                                                       
5       Marly SPF S.A.                                                                                                                       
6       I.T APPARELS LIMITED                                                                                                                 
7       VX 30.141 ApS                                                                                                                        
8       Rydex ETF Trust - Guggenheim S&P 500 Equal Weight ETF                                                                                
9       Rydex Series Funds - Retailing Fund                                                                                                  
10      Olam Holdings, Inc.                                                                                                                  
11      Fire Holding A/S                                                                                                                     
12      Aarhus Murerlaug                                                                                                                     
13      Indigo Hall SC JV Investors, LLC                                                                                                     
14      Hestbech Holding ApS                                                                                                                 
15      Pharmatrade A/S                                                                                                                      
16      GS+A Tactical Fixed Income Master Fund (Cayman) LP                                                                                   
17      Lars Nielsens Fond                                                             

In [30]:
companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 450256 entries, 0 to 450255
Data columns (total 4 columns):
company_id    450256 non-null int64
name          450256 non-null object
clean_name    450256 non-null object
n_grams       450256 non-null object
dtypes: int64(1), object(3)
memory usage: 13.7+ MB


In [31]:
train["n_grams"] = make_corpus(train["clean_name"], N_GRAMS)

In [32]:
train

train_index  \
0      0             
1      1             
2      2             
3      3             
4      4             
5      5             
6      6             
7      7             
8      8             
9      9             
10     10            
11     11            
12     12            
13     13            
14     14            
15     15            
16     16            
17     17            
18     18            
19     19            
20     20            
21     21            
22     22            
23     23            
24     24            
25     25            
26     26            
27     27            
28     28            
29     29            
...    ..            
99970  99970         
99971  99971         
99972  99972         
99973  99973         
99974  99974         
99975  99975         
99976  99976         
99977  99977         
99978  99978         
99979  99979         
99980  99980         
99981  99981         
99982  99982         
99983  99983         
99984  99984         
99985  99985         
99986  99986         
99987  99987         
99988  99988         
99989  99989         
99990  99990         
99991  99991         
99992  99992         
99993  99993         
99994  99994         
99995  99995         
99996  99996         
99997  99997         
99998  99998         
99999  99999         

                                                                                      name  \
0      TRATTAMENTO Ltd RIFIUTI METROPOLITANI SPA  SIGLABILE TRM SPA                          
1      A IRL Fuund                                                                           
2      BMR-500 Kendall LLC 1 Mezz GmbH                                                       
3      Solich GmbH KG                                                                        
4      Drzyzzga Funds Logi. sp. z oo                                                         
5      BERK ELEY LLP                                                                         
6      SCHNEIDER ELECTRIC PROTECTION & CONTROLE                                              
7      DL ISL Inc                                                                            
8      Arendicom GmbH                                                                        
9      RBPA Leeuw                                                                            
10     NORP. Holdingproteccion PPS-NORP. Holding 2015                                        
11     Realty, LLC                                                                           
12     CP X, spol. s r.o.                                                                    
13     THE FIONA RUTH HAWKER 2009 DISC RETIONARY SETTLEMENT                                  
14     Dansk Socialrådgiverforening                                                          
15     Elek. zReszów SA                                                                      
16     SCAN LUX PACKA. SICAV A/S                                                             
17     Lars Väste. i Hammar AB                                                               
18     9208-2932 Inc Québec.                                                                 
19     Waddell & Reed Advisors Funds - Waddell & Reed Advisors Small Cap Fund                
20     HBOS LIMITED INTERNATIONAL FS HOLDINGS LIMITED                                        
21     Bravofood Sytsems, LLC                                                                
22     ESCO TORE S.R.L.                                                                      
23     Columbia Funds Series Trust II - Columbia Portfolio - Select Smaller-aCp Value Fund   
24     LObCA Servi. SL                                                                       
25     GIOVANNI S.A.S. DI DONISOTTI DOPRA & C.                                               
26     Sand alos2001 SICAV, S.A.                                                             
27     THE

### Prepare training data for learning embeddings.
*  Word2vec learns embeddings based on the cotext of words in corpus. Starspace can be trained to learn  embeddings (word, images etc.) based on specific task. Thus it can learn better embeddings. In our case the task is name similarity.
* Starspace training format needs pairs of similar setences. So we use names from the train set with a m,atching company id in companies set
* <b>Hack</b>: Training set has only a fraction of all the company names. For company names not present in train set we add make a pair using the true company name.

In [33]:
train_data = companies.merge(train[["clean_name", "n_grams","company_id"]], how="left", on="company_id", suffixes=('','_train'))

In [34]:
train_data["company_id"].value_counts()

32434     5
425226    5
165777    4
634977    4
407946    4
440786    4
88538     4
209368    4
177140    4
295600    4
638323    3
610823    3
207022    3
172728    3
269318    3
618417    3
198931    3
110194    3
204786    3
365550    3
306655    3
499321    3
603660    3
595973    3
530204    3
170350    3
328231    3
196758    3
224132    3
186928    3
         ..
80243     1
65908     1
55631     1
51533     1
250158    1
49484     1
252207    1
205104    1
209202    1
211251    1
196916    1
203063    1
223545    1
227643    1
213308    1
215357    1
217406    1
41280     1
43329     1
45378     1
47427     1
33092     1
35141     1
37190     1
39239     1
57672     1
59721     1
61770     1
63819     1
0         1
Name: company_id, Length: 450256, dtype: int64

In [35]:
train_data["clean_name_train"] = train_data["clean_name_train"].fillna(train_data["clean_name"])
train_data["n_grams_train"] = train_data["n_grams_train"].fillna(train_data["n_grams"])

In [36]:
train_data.head()

company_id                        name                  clean_name  \
0  634022      PRIMCOM SA                  primcom sa                   
1  324497      The David Isaacs Fund       the david isaacs fund        
2  280848      Bramor Enterprises Limited  bramor enterprises limited   
3  432662      NAVEXIM S.A.                navexim sa                   
4  524224      Magal Group SA              magal group sa               

                                                                                           n_grams  \
0  pri rim imc mco com om_ m_s _sa                                                                   
1  the he_ e_d _da dav avi vid id_ d_i _is isa saa aac acs cs_ s_f _fu fun und                       
2  bra ram amo mor or_ r_e _en ent nte ter erp rpr pri ris ise ses es_ s_l _li lim imi mit ite ted   
3  nav ave vex exi xim im_ m_s _sa                                                                   
4  mag aga gal al_ l_g _gr gro rou oup up_ p_s _sa                                                   

                         clean_name_train  \
0  primcom sa                               
1  the david isaacs fund                    
2  bramor inves tment enterprises limited   
3  navexim sa                               
4  magal group sa                           

                                                                                                                                     n_grams_train  
0  pri rim imc mco com om_ m_s _sa                                                                                                                  
1  the he_ e_d _da dav avi vid id_ d_i _is isa saa aac acs cs_ s_f _fu fun und                                                                      
2  bra ram amo mor or_ r_i _in inv nve ves es_ s_t _tm tme men ent nt_ t_e _en ent nte ter erp rpr pri ris ise ses es_ s_l _li lim imi mit ite ted  
3  nav ave vex exi xim im_ m_s _sa                                                                                                                  
4  mag aga gal al_ l_g _gr gro rou oup up_ p_s _sa

In [37]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 455388 entries, 0 to 455387
Data columns (total 6 columns):
company_id          455388 non-null int64
name                455388 non-null object
clean_name          455388 non-null object
n_grams             455388 non-null object
clean_name_train    455388 non-null object
n_grams_train       455388 non-null object
dtypes: int64(1), object(5)
memory usage: 24.3+ MB


In [38]:
#train_data = train_data[["clean_name", "clean_name_train"]]
train_data = train_data[["n_grams", "n_grams_train"]]

In [39]:
train_data.to_csv(file_path+"star_train_ngrams.tsv",sep="\t",header=False)

### Learn Embeddings

In [40]:
#/home/akash/projects/Starspace/starspace train -trainFile star_train_ngrams.tsv -model star_space_ngram_embeddings -trainMode 3 -adagrad true -ngrams 1 -epoch 5 -dim 100 -similarity "cosine" -minCount 2 -verbose true -fileFormat labelDoc -negSearchLimit 10 -lr 0.05 -thread 4

#### Build star space embeddings for ngrams

In [41]:
star_space_dict = {}
with open(file_path+"/star_space_ngram_embeddings.tsv") as f:
    for line in f:
        line = line.split("\t")
        token = line[0]
        embedding = line[1:]
        embedding = np.asarray(embedding).astype(np.float32)     
        star_space_dict[token] = embedding      
starspace_embeddings = star_space_dict

In [42]:
print (starspace_embeddings['pri'])

[-0.119208    0.0991629   0.0938104   0.0660154   0.0747938   0.070145
  0.130213   -0.00070346  0.0130067  -0.0293467   0.0686699   0.0810226
 -0.012915   -0.0572339  -0.0317828   0.0658     -0.0950785   0.113955
  0.0173498   0.0279843  -0.0506483   0.125173   -0.0619245  -0.0726998
  0.0572543  -0.0402238  -0.0666396  -0.152028    0.0115009   0.121052
  0.0560702   0.0278294  -0.0976862  -0.026822   -0.00935354 -0.0404374
 -0.0716217   0.167094   -0.0235991   0.109166   -0.0032483   0.0235811
 -0.0627361  -0.0391092  -0.00491249 -0.102261   -0.0518779   0.0241284
 -0.032624   -0.125606    0.120422    0.101578   -0.0462269   0.0371669
 -0.0469588   0.103301   -0.0593593   0.0787304  -0.0110581   0.104398
 -0.0594812  -0.0300357  -0.119987    0.0919375   0.00581149  0.0338233
 -0.076168    0.0405324   0.0383356  -0.0534815  -0.038259    0.0862937
 -0.107805    0.0741736  -0.0283083  -0.0437005  -0.0513222  -0.00458159
  0.116911   -0.0208172   0.0741671   0.016341   -0.0341246  -0.077

In [43]:
def name_to_vec(name, embeddings):
    """
        name: a company name/sentence
        embeddings: dict where the key is a word and a value is its' embedding
        dim: size of the representation

        result: vector representation for the name/sentence
    """
    
    result = np.zeros(shape=DIM, dtype=float)
    count = 0
    #print name
    for word in name.split():
        #print word
        if word in embeddings:
            #print "word %s in embedding"%word
            result = result + embeddings[word]
            count+=1
    if count>0:
        result = result/count
        
    return result

In [44]:
def make_name_vectors(corpus, embeddings):
    name_vectors = np.empty(shape=(len(corpus),DIM))
    for idx, name in enumerate(corpus):
        name_vector = name_to_vec(name, embeddings)
        name_vectors[idx,:] = name_vector
    return name_vectors

In [45]:
len(starspace_embeddings)

22004

#### Make embedding vectors of company names by taking the mean of ngram embeddings

In [46]:
company_vectors = make_name_vectors(companies["n_grams"], starspace_embeddings)

In [47]:
company_vectors[0:1]

array([[ 0.01709783,  0.00773632, -0.0021093 ,  0.01139055, -0.02135037,
         0.0146049 ,  0.03187227,  0.01187479,  0.00035386, -0.02018384,
        -0.00339857,  0.03907295, -0.01720327, -0.02469894,  0.02020706,
         0.00131685, -0.06940149,  0.07446843, -0.00120021,  0.00548082,
        -0.01596069, -0.00300824, -0.01614329, -0.04761717,  0.04952533,
        -0.02928773,  0.01583691, -0.03834475, -0.00712968,  0.01294641,
        -0.02833813, -0.00174684,  0.0097028 , -0.00415216,  0.01987893,
        -0.01534769,  0.03831299,  0.06816379,  0.0051133 ,  0.06606996,
        -0.00863443, -0.00949614, -0.01702975, -0.02929292, -0.01829027,
        -0.06338784, -0.0129171 , -0.04590585,  0.02071874, -0.03460466,
         0.04835812,  0.04926606,  0.01057489,  0.04920899,  0.02093529,
        -0.02822798,  0.02639537,  0.02562038, -0.0001477 ,  0.0099464 ,
        -0.00158709, -0.04102737, -0.01340589, -0.01744403, -0.01889571,
         0.0188464 , -0.04703097,  0.01120369, -0.0

### Indexing using Annoy.
Approximate nearest neigbour using random projections. Indexes once built are static files and can be used across processes.

#### Make map of company_ids to annoy_index_ids

In [48]:
# annoy_ids are from 0 to len(company_ids) -1 
# annoy_ids are stored in list. List index are annoy_ids and the value is teh correspodnign company_id
# Company_ids are stored in a dict. The key is company id and the value is annoy_id
def build_company_annoy_maps(company_ids):
    annoy2company = []
    company2annoy = defaultdict(lambda:0)
    for c_id in company_ids:
        if c_id not in company2annoy:
            annoy2company.append(c_id)
            company2annoy[c_id] = len(annoy2company)-1
    return annoy2company, company2annoy  
        

In [49]:
annoy2company, company2annoy = build_company_annoy_maps(companies["company_id"].values)

In [50]:
print len(annoy2company)
print len(company2annoy)

450256
450256


In [51]:
import random
for i in range(5):
    annoy_id = random.randint(0,50000)
    assert annoy_id ==  company2annoy[annoy2company[annoy_id]]
    print annoy_id, annoy2company[annoy_id], company2annoy[annoy2company[annoy_id]]

34100 198710 34100
4580 484544 4580
30891 320028 30891
42096 89805 42096
41728 232717 41728


#### Build annoy Index
Uncoment below to build the index

In [52]:
# index_size = DIM
# table = AnnoyIndex(index_size)

# for i in range(company_vectors.shape[0]):
#     if i%20000==0:
#         print "indexed %s items"%i
#     table.add_item(i,company_vectors[i])


#table.build(NUM_TREES_ANNOY)
#table.save(file_path+"annoy_starspace_ngram_index.ann")

###  Load Index and search

In [53]:
index_size = DIM
table = AnnoyIndex(index_size)
table.load(file_path+"annoy_starspace_ngram_index.ann")

True

###  Find neigbhours of train set


#### Build training name vectors

In [54]:
# function to fetch nearest neighbours
def get_nearest_neighbours(vectors, num_neighbours=1, search_nodes=-1, include_distances=True):
    neighbours = np.empty(shape=(len(vectors),num_neighbours), dtype=np.int32)
    distances = np.empty(shape=(len(vectors),num_neighbours))
#     print len(vectors)
#     print neighbours.shape
#     print distances.shape
    for idx,v in enumerate(vectors):
        annoy_ids, annoy_distances = table.get_nns_by_vector(v, n=num_neighbours, search_k=search_nodes, include_distances=include_distances)
        neighbours[idx,:] = annoy_ids
        distances[idx,:] = annoy_distances
    return neighbours, distances

#### We can retrieve n nearest neighborus, but since we need find only the ground truth I fetch only the nearest neighbour 


In [55]:
train.shape

(100000, 5)

In [56]:
%%time
train_vectors = make_name_vectors(train["n_grams"],starspace_embeddings)
neighbours, distances = get_nearest_neighbours(train_vectors)
neighbour_company_ids =np.vectorize(lambda x: annoy2company[x])(neighbours)

CPU times: user 26.4 s, sys: 35.7 ms, total: 26.5 s
Wall time: 26.5 s


#### Set nearest neighbour to be the prediction

In [57]:
rank_1 = neighbour_company_ids[:,0]
train["predicted_id"] = rank_1

In [58]:
train= train.merge(companies[['company_id', 'name']], left_on="predicted_id", 
                   right_on="company_id", suffixes=("","_predicted"))

In [59]:
train.head()

train_index                                                          name  \
0  0            TRATTAMENTO Ltd RIFIUTI METROPOLITANI SPA  SIGLABILE TRM SPA   
1  1            A IRL Fuund                                                    
2  2            BMR-500 Kendall LLC 1 Mezz GmbH                                
3  3            Solich GmbH KG                                                 
4  4            Drzyzzga Funds Logi. sp. z oo                                  

   company_id                                                   clean_name  \
0  177358      trattamento ltd rifiuti metropolitani spa siglabile trm spa   
1  568472      a irl fuund                                                   
2  195692      bmr 500 kendall llc 1 mezz gmbh                               
3 -1           solich gmbh kg                                                
4  404178      drzyzzga funds logi sp z oo                                   

                                                                                                                                                                                                                               n_grams  \
0  tra rat att tta tam ame men ent nto to_ o_l _lt ltd td_ d_r _ri rif ifi fiu iut uti ti_ i_m _me met etr tro rop opo pol oli lit ita tan ani ni_ i_s _sp spa pa_ a_s _si sig igl gla lab abi bil ile le_ e_t _tr trm rm_ m_s _sp spa   
1  a_i _ir irl rl_ l_f _fu fuu uun und                                                                                                                                                                                                   
2  bmr mr_ r_5 _50 500 00_ 0_k _ke ken end nda dal all ll_ l_l _ll llc lc_ c_1 _1_ 1_m _me mez ezz zz_ z_g _gm gmb mbh                                                                                                                   
3  sol oli lic ich ch_ h_g _gm gmb mbh bh_ h_k _kg                                                                                                                                                                                       
4  drz rzy zyz yzz zzg zga ga_ a_f _fu fun und nds ds_ s_l _lo log ogi gi_ i_s _sp sp_ p_z _z_ z_o _oo                                                                                                                                   

   predicted_id  company_id_predicted  \
0  177358        177358                 
1  223012        223012                 
2  195692        195692                 
3  497655        497655                 
4  331552        331552                 

                                                  name_predicted  
0  TRATTAMENTO RIFIUTI METROPOLITANI S.P.A. SIGLABILE TRM S.P.A.  
1  IMPRESA EDILE CAMBREA ROCCO S.R.L                              
2  BRE-BMR-500 Kendall Mezz 1 LLC                                 
3  Fröhlich GmbH                                                  
4  SITS Sp. z o.o.

In [60]:
train[["train_index", "name", "company_id", "predicted_id", "name_predicted"]]

train_index  \
0      0             
1      1             
2      2             
3      3             
4      4             
5      36093         
6      5             
7      6             
8      7             
9      8             
10     41918         
11     9             
12     10            
13     99945         
14     11            
15     73484         
16     76226         
17     82038         
18     12            
19     93356         
20     13            
21     81035         
22     14            
23     15            
24     60665         
25     16            
26     17            
27     18            
28     19            
29     20            
...    ..            
99970  99956         
99971  99957         
99972  99958         
99973  99959         
99974  99960         
99975  99962         
99976  99963         
99977  99965         
99978  99966         
99979  99967         
99980  99970         
99981  99972         
99982  99973         
99983  99974         
99984  99976         
99985  99978         
99986  99979         
99987  99981         
99988  99982         
99989  99984         
99990  99985         
99991  99986         
99992  99987         
99993  99988         
99994  99989         
99995  99991         
99996  99992         
99997  99993         
99998  99996         
99999  99997         

                                                                                               name  \
0      TRATTAMENTO Ltd RIFIUTI METROPOLITANI SPA  SIGLABILE TRM SPA                                   
1      A IRL Fuund                                                                                    
2      BMR-500 Kendall LLC 1 Mezz GmbH                                                                
3      Solich GmbH KG                                                                                 
4      Drzyzzga Funds Logi. sp. z oo                                                                  
5      Aurepio TRUST Sp. z o.o.                                                                       
6      BERK ELEY LLP                                                                                  
7      SCHNEIDER ELECTRIC PROTECTION & CONTROLE                                                       
8      DL ISL Inc                                                                                     
9      Arendicom GmbH                                                                                 
10     Deltgcenterscape 5 GmbH Holding                                                                
11     RBPA Leeuw                                                                                     
12     NORP. Holdingproteccion PPS-NORP. Holding 2015                                                 
13     ungi. Corpo rationpte. Ltd.                                                                    
14     Realty, LLC                                                                                    
15     Realty, LLC                                                                                    
16     Realty, LLC                                                                                    
17     ADGA Realty, LLC                                                                               
18     CP X, spol. s r.o.                                                                             
19     ORNEX, spol. s r.o.                                                                            
20     THE FIONA RUTH HAWKER 2009 DISC RETIONARY SETTLEMENT                                           
21     THE C V GEOGHEGAN 2009 DISCRETIONARY SETT.                                                     
22     Dansk Socialrådgiverforening                                                                   
23     Elek. zReszów SA                                                                               
24     Elektromontż Rzeszów SA                                     

### Evaluate
#### Since we do not handle -1 cases we only report accuracy.

In [61]:
total = len(train)
correct = train.loc[train["company_id"]==train["predicted_id"]].shape[0]
incorrect = train.loc[train["company_id"]!=train["predicted_id"]].shape[0]
print "Total: %d"%total
print "Correct Predictions %f"%(correct/float(total))
print "Incorrect Predictions %f"%(incorrect/float(total))

Total: 100000
Correct Predictions 0.372670
Incorrect Predictions 0.627330


### On excluding -1 cases accuracy is 53.4%

In [62]:
train_sub = train.loc[train["company_id"]!=-1]
print np.sum(train_sub["company_id"] == train_sub["predicted_id"])/float(len(train_sub))

0.5343398715301675


### Accuracy = 0.5343
* n_grams=[3] 
* dim =100
* num_trees_annoy=200

Total: 100000
Correct Predictions 0.372670
Incorrect Predictions 0.627330




### Generate Test predictions

In [63]:
test = pd.read_csv(data_path+"STest.csv", delimiter=",")
test.tail()

test_index                                                       name
99995  99995       Spó�jama�ka z ograniczon� Investmexnt� odpowiedzialnością
99996  99996       Banbrico LLC Limited                                     
99997  99997       The NF                                                   
99998  99998       OLZ Holding - EWD Equity (ex CH)                         
99999  99999       VÆ299 99-NGE

In [64]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 2 columns):
test_index    100000 non-null int64
name          99999 non-null object
dtypes: int64(1), object(1)
memory usage: 1.5+ MB


In [65]:
test.isnull().sum()

test_index    0
name          1
dtype: int64

In [66]:
#### Test set has missing names as well as incorrectly encoded names. We drp these

In [67]:
test = test.loc[test["name"].notnull()]

In [68]:
test = name_pipeline.transform(test)

None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None


AttributeError: 'NoneType' object has no attribute 'split'

In [69]:
test.isnull().sum()

test_index    0   
name          0   
clean_name    2138
dtype: int64

In [70]:
test.loc[test["clean_name"].isnull()]

test_index  \
120    120          
124    124          
184    184          
191    191          
208    208          
232    232          
237    237          
248    248          
283    283          
320    320          
388    388          
424    424          
466    466          
478    478          
753    753          
1107   1107         
1125   1125         
1128   1128         
1131   1131         
1163   1163         
1230   1230         
1267   1267         
1319   1319         
1329   1329         
1344   1344         
1380   1380         
1458   1458         
1520   1520         
1549   1549         
1573   1573         
...     ...         
98369  98369        
98410  98410        
98464  98464        
98570  98570        
98599  98599        
98674  98674        
98695  98695        
98864  98864        
98925  98925        
98938  98938        
98963  98963        
99055  99055        
99071  99071        
99124  99124        
99160  99160        
99273  99273        
99342  99342        
99434  99434        
99443  99443        
99444  99444        
99478  99478        
99563  99563        
99693  99693        
99741  99741        
99747  99747        
99810  99810        
99914  99914        
99956  99956        
99958  99958        
99995  99995        

                                                                                                    name  \
120    ASSISTEC SOCIETATE CU RL�                                                                           
124    RAMEXA ll� Hold ings TermékelőáIpari�s Zárűen működő rétKrs�nytársaság                              
184    WARSAW GAS TRADING SPÓŁKA Z OGRAN.� ODPOWIEDZIALNOŚCIĄ                                              
191    Eurogalva - Galvanização e Metal.�nia S.A.                                                          
208    Chemgas Schiffahrts UG (haftungsbeschr�nkt) & Co. MT "GASCHEM RHONE" KG                             
232    "M" S FÉRMMR� KORLÁTOLT FELELŐSSÉGŰ TÁRSASÁG                                                        
237    Harald FUND Ullman Förs�k ringsrätt AB                                                              
248    Arionn Sp. z o.o. 3 Spó� Fund�ka koman dytowo-akcyjna                                               
283    MA rs� & FöKonsult�ljning Equity AB                                                                 
320    Budap. Alap�r Rövid Kötvény Doll�                                                                   
388    FUND.�N Y AE RODADO LEAL                                                                            
424    Atos teri̇li̇şi̇m Danişmanlik Ve Müşfi̇ Hi� FUND�zmet leri̇ Sanayi̇ Ve Ti̇caret Anoni̇m Şi̇rketi̇   
466    Fru Limited E. Hj� Legat For Syge, Fattige Og Scha.�lpeløse Perso ner                               
478    W&N Bautr�ger GmbH                                                                                  
753    Stone gu TáKtr�g                                                                                    
1107   Jalo Funds Verm�gensverwaltungs KG                                                                  
1125   AFI Verwa ltungs-Gese. mb�nkter Haftung                                                             
1128   JKH s KSlK Holdings�felelősségű társaság                                                            
1131   USV FUND�gens. KG                                                                                   
1163   Premier G. Med Kereskedelmi Korlátolt Felelősségű Társas� Funds�g                                   
1230   Praktiserende mbH Land inspekt�rers Foren.                                                          
1267   ALFÖLD Porcelán Ed�ny gy� Trust�r Zrt.                                                              
1319   ka Z ESOO LLC�                                                                                      
1329   Ślusarstwo Usługowe Hajduk ka z ograniczoną odpowiedzialnością komandytowa�ka spó� SICAV            
1344

In [71]:
test = test.loc[test["clean_name"].notnull()]

In [72]:
test["n_grams"] = make_corpus(test["clean_name"], N_GRAMS)

/home/akash/.local/lib/python2.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [73]:
%%time
test_vectors = make_name_vectors(test["n_grams"],starspace_embeddings)
test_neighbours, test_distances = get_nearest_neighbours(test_vectors)
test_neigbours =np.vectorize(lambda x: annoy2company[x])(test_neighbours)

CPU times: user 25.6 s, sys: 43.9 ms, total: 25.7 s
Wall time: 25.7 s


In [74]:
print test_vectors.shape

(97861, 100)


In [75]:
print test_neighbours

[[ 78025]
 [445548]
 [174579]
 ...
 [ 82655]
 [370937]
 [ 62907]]


#### Set nearest neighbour to be the prediction

In [76]:
test["predicted_id"] = test_neigbours[:,0]

/home/akash/.local/lib/python2.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [77]:
test= test.merge(companies[['company_id', 'name']], left_on="predicted_id", 
                   right_on="company_id", suffixes=("","_predicted"))

In [78]:
test

test_index  \
0      0            
1      1            
2      2            
3      12512        
4      3            
5      21903        
6      4            
7      5            
8      6            
9      7            
10     8            
11     9            
12     10           
13     16219        
14     30914        
15     99986        
16     11           
17     63186        
18     12           
19     25605        
20     13           
21     14           
22     15           
23     16           
24     17           
25     38484        
26     18           
27     19           
28     20           
29     21           
...    ..           
97831  99959        
97832  99960        
97833  99961        
97834  99963        
97835  99964        
97836  99965        
97837  99967        
97838  99969        
97839  99970        
97840  99971        
97841  99972        
97842  99973        
97843  99974        
97844  99975        
97845  99976        
97846  99977        
97847  99978        
97848  99980        
97849  99981        
97850  99982        
97851  99983        
97852  99984        
97853  99987        
97854  99988        
97855  99991        
97856  99992        
97857  99993        
97858  99996        
97859  99997        
97860  99999        

                                                                                                                                           name  \
0      THEking'S ROYAL HUSSARS OFFI. TRUST' TRUST                                                                                                 
1      Southern Powe rcompany SICAV                                                                                                               
2      BMO S&P/TSX Ladde. Share ETF Index                                                                                                         
3      ProShares S&P 500 ex-Health Care ETF                                                                                                       
4      PaI                                                                                                                                        
5      uNTL                                                                                                                                       
6      Clearview Two                                                                                                                              
7      CARTERA OkPTIMA MODERADA, FI                                                                                                               
8      HOBART 85 LIMITED GmbH                                                                                                                     
9      SOC MAX MODEL                                                                                                                              
10     Bill Kropa.                                                                                                                                
11     FuG & Co KG                                                                                                                                
12     Icep atrimonial SPRL                                                                                                                       
13     SOCIETA PRO APRMA IMMO BILIARE' PA                                                                                                         
14     Ice Patrimonialsprl                                                                                                                        
15     COSTANZO FRAT. UMBERTO E PAOLO                                                                                                             
16     TUS Limited Leland                                                                                                                         
17     Hella Zealand Limited                              

#### Save test predictions

In [79]:
test[["test_index", "company_id"]].to_csv(data_path+"test_predictions.csv", sep="|", index=False)

In [80]:
test.shape

(97861, 7)

That is all!